### PR TITLE
irmin-pack: introduce Sparse_file abstraction

### PR DESCRIPTION
**Pull request by art-w**
This PR tries to abstract the `mapping/data` files coupling behind a `Sparse_file` module (that we could use to implement the lower volumes?). It's a pretty aggressive refactoring so I would understand if it is not welcomed! It notably gets rid of the dispatcher's accessors because I find it easier to hide the offset computations inside the Sparse/Chunked files blackboxes.

If there's no strong objection, my plan is to add some append-only functions to this `Sparse_file` that the GC will be able to use to transfer data to the lower. There's also still some refactoring to do in `File_manager` as it could delegate the opening/closing/etc of the mapping/data file to the new module.

### CODE REVIEW

File: src/irmin-pack/unix/sparse_file.ml [hunk 1]
```diff
@@ -0,0 +1,335 @@
...
+          src.{i + 1} <- poff;
+          let len = src.{i + 2} in
+          go (i + 3) (poff + len))
+      in
+      go 0 0
+
+    type t = { arr : int64_bigarray; root : string; generation : int }
+
+    let open_map ~root ~generation =
+      let path = Irmin_pack.Layout.V4.mapping ~generation ~root in
```
Comment (metanivek) [inline src/irmin-pack/unix/sparse_file.ml:135]: Are you planning to update these hard-coded paths in this PR or a future one?

File: src/irmin-pack/unix/sparse_file_intf.ml [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        The exceptions raised by [f] are caught and returned (as long as they
+        are known by [Errs]). *)
+
+    val iter_exn : t -> (off:int63 -> len:int -> unit) -> unit
+    (** Similar to [iter mapping f] but raises exceptions. *)
+  end
+
+  type t
+
+  val make : mapping:Mapping_file.t -> data:Io.t -> t
```
Comment (metanivek) [inline src/irmin-pack/unix/sparse_file_intf.ml:71]: Should we stick with the `v` naming convention?

File: src/irmin-pack/unix/utils.ml [hunk 3]
```diff
@@ -63,22 +63,24 @@ end = struct
...
-    |> iter_k (fun ~k (lo, hi) ->
-           if lo + 1 = hi then `Some lo
-           else
-             let mid = (lo + hi) / 2 in
-             if get arr mid <= key then k (interval mid hi)
-             else k (interval lo mid))
+(** [nearest_geq ~arr ~get ~lo ~hi ~key] returns the index of the nearest entry
+    in the sorted [arr] that is [>=] the given key. Routine is based on binary
+    search. *)
+let nearest_geq ~arr ~get ~lo ~hi ~key =
```
Comment (metanivek) [inline src/irmin-pack/unix/utils.ml:69]: Why did you change from `leq` to `geq`?

**Review by metanivek — COMMENTED** [review]
I've only looked at the core of the changes so far but will do a more thorough review later.

Overall, this does seem to make the reading code more direct, and I like packaging the logic of the sparse file concept into a separate module since it will be useful for volumes too.

I originally thought we would extend accessors to include location information when reading from volumes, but I think it won't be so bad without them. The way we will have to do it without accessors is to change the return type of `read_exn` from `unit` to some kind of location information.

File: src/irmin-pack/unix/dispatcher.ml [hunk 4]
```diff
@@ -77,309 +63,85 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+    let open Int63.Syntax in
+    let suffix_start_offset = suffix_start_offset t in
+    let suffix_dead_bytes = suffix_dead_bytes t in
+    suffix_start_offset + soff - suffix_dead_bytes
 
   let end_offset t =
     let end_soff = Suffix.end_soff (Fm.suffix t.fm) in
     offset_of_soff t end_soff
 
-  module Prefix_arithmetic = struct
```
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/dispatcher.ml:109]: I believe it's the same logic as before, with slightly less verbose error messages (but I did remove some dead code that was checking for the correctness of the `leq` function). The functions `geq` and `leq` are very similar in behavior and edge-cases to handle so I don't think the change helped much in that respect.

File: src/irmin-pack/unix/dispatcher.ml [hunk 5]
```diff
@@ -77,309 +63,85 @@ module Make (Fm : File_manager.S with module Io = Io.Unix) :
...
+  let read_range_exn t ~off ~min_len ~max_len buf =
+    match dispatch_suffix t ~off with
+    | Some off ->
+        Suffix.read_range_exn (Fm.suffix t.fm) ~off ~min_len ~max_len buf
+    | None -> Sparse.read_range_exn (get_prefix t) ~off ~min_len ~max_len buf
 
   let read_bytes_exn t ~f ~off ~len =
-    let open Int63.Syntax in
     let bytes_in_prefix =
+      let open Int63.Syntax in
```
[the remark this answers is not in the document]
Comment (art-w) [reply src/irmin-pack/unix/dispatcher.ml:102]: I actually didn't have much choice as there are `int` operations below :P

File: src/irmin-pack/unix/sparse_file.ml (hunk 1 above)
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml]: Oh thanks I missed it when shuffling the code around!

File: src/irmin-pack/unix/utils.ml (hunk 3 above)
Comment (art-w) [reply src/irmin-pack/unix/utils.ml:69]: I needed to support traversing the sparse file from left to right (for `Traverse_pack_file` and `irmin-tezos-utils`), jumping from one "object" `(off,len)` to the next `(off+len,...)` but skipping over the sparse holes... so I first added `geq` then realized that `leq` was redundant :) (`geq` gives you the next valid offset, while `leq` would have worked for a right-to-left traversal)

**Comment by art-w** [comment]
Thanks for the review! I've made some changes (to be squashed after approval) and simplified a bit the file manager (... I'm trying to expose less and less of the `Mapping_file`, although the GC worker still need this internal access until the sparse file provides "append" operations)

File: src/irmin-pack/unix/dispatcher.ml (hunk 5 above)
[the remark this answers is not in the document]
Comment (metanivek) [reply src/irmin-pack/unix/dispatcher.ml:102]: Ah yes, but you moved it from opened in the entire function to just the places you needed it (for `bytes_in_prefix` and `bytes_in_suffix`) and then using `Int63.Syntax` directly at the end. I'm just a fan of opening in smaller scopes. :smile:

File: src/irmin-pack/unix/utils.ml (hunk 3 above)
Comment (metanivek) [reply src/irmin-pack/unix/utils.ml:69]: Ah, okay. That history makes sense!

File: src/irmin-pack/unix/file_manager.ml [hunk 6]
```diff
@@ -206,7 +206,8 @@ struct
     let open Result_syntax in
     if generation = 0 then Ok None
     else
-      let* m = Mapping_file.open_map ~root ~generation in
+      let path = Irmin_pack.Layout.V3.mapping ~generation ~root in
```
Comment (metanivek) [inline src/irmin-pack/unix/file_manager.ml]: All these should use `V4`, just for consistency.

(Future PR: I think we ought to have a `Latest` layout, similar to what we do for the control file, to make these a bit less of a burden to maintain)

File: src/irmin-pack/unix/sparse_file.ml [hunk 7]
```diff
@@ -297,7 +293,7 @@ module Make (Io : Io.S) = struct
   let get_poff { mapping; _ } ~off =
     match Mapping_file.find_nearest_geq mapping off with
     | None ->
-        let s = Fmt.str "offset %a is after the prefix" Int63.pp off in
+        let s = Fmt.str "offset %a is after the sparse file" Int63.pp off in
```
Comment (metanivek) [inline src/irmin-pack/unix/sparse_file.ml:303]: Since this error could refer to different files once we have volumes, it would be nice to include the path to either the mapping or data file in this error message.

Also, the error `Invalid_read_of_gced_object` might deserve a rename. Perhaps, `Invalid_read_of_sparse_object` (or something like this that is more generic).

File: src/irmin-pack/unix/sparse_file.ml [hunk 8]
```diff
@@ -285,6 +285,13 @@ module Make (Io : Io.S) = struct
   type t = { mapping : Mapping_file.t; data : Io.t }
 
   let v ~mapping ~data = { mapping; data }
+
+  let open_ro ~mapping ~data =
+    let open Result_syntax in
+    let* mapping = Mapping_file.open_map ~path:mapping in
+    let+ data = Io.open_ ~path:data ~readonly:true in
+    { mapping; data }
```
Comment (metanivek) [inline src/irmin-pack/unix/sparse_file.ml:293]: I view `v` as the canonical constructor of the type so would prefer this to be `v ~mapping ~data`.

**Review by metanivek — COMMENTED** [review]
Really like the cleanups! Left some new comments for consideration.

File: src/irmin-pack/unix/sparse_file.ml (hunk 7 above)
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml:303]: Yeah I agree that it would be nicer if all errors had the filenames for debugging user issues. Do you mind if we do a separate PR to improve the error reporting?
(I did switch the `Invalid_read_of_gced_object` for `Invalid_sparse_read` to match with the other existing errors)

File: src/irmin-pack/unix/sparse_file.ml (hunk 8 above)
Comment (art-w) [reply src/irmin-pack/unix/sparse_file.ml]: Ha thanks, I missed it! (but the `v` is hopefully going away soon :-°)

File: src/irmin-pack/unix/sparse_file.ml (hunk 7 above)
Comment (metanivek) [reply src/irmin-pack/unix/sparse_file.ml:303]: Yep, totally fine with a follow up PR to improve errors!

**Review by metanivek — APPROVED** [review]
LGTM!
